### PR TITLE
Creating Sign In Command

### PR DESCRIPTION
**Pull request by samjonester**
Closes rubygems/rubygems#1634

Sign in command signs the user into a gem's host prior to pushing a gem.
Uses the same sign in mechanism as `PushCommand`.
Can be used to sign into a gem's host prior to running automated tooling
to push gem.

### CODE REVIEW

**Comment by tenderlove** [comment]
This seems like a nice feature.  Should it have tests though?

**Comment by samjonester** [comment]
Nah, testing is for schmuks! Jk! I :green_heart: testing. I work for Test Double!

I wanted to gauge interest in this feature, before I invested too much time in it. There was a closed feature request for this functionality previously (#779). That was a while ago, though, and I believe I've got a better case for its support. 

1. Automation is more prevalent, so `gem push` can't be assumed to be run manually, anymore. It may very well be a part of a CI workflow. 
2. Bundler includes `gem push` in its packaged task `rake release`.
3. This benefits more than just gemcutter. My use case was artifactory. 

If you're comfortable with the API, there's interest in merging this down, I'd be happy to shore it up with tests. I'd also like to:
* do some cleanup around authentication in the `push` command and it's tests. 
* consider providing sign in functionality through a dependency rather than inheritance. This would allow better and more complete testing of authentication for both `push` and `signin`.
* take a stab at adding documentation for the new command.

**Comment by samjonester** [comment]
I've implemented tests for the sign in command, and in a separate commit (that can potentially be dropped), I've renamed Gem Cutter Utilities to Authorization Utilities. They are not specific to Gem Cutter, but instead used for the more generic authorization actions.

Where should the documentation for this new command be updated?

Additionally, there are some issues with the push command that should still be addressed. Should I include this as a separate commit in this PR?
1. There are some tests in the push command that pass when setup is deleted, leading me to think they aren't testing what they think they are.
2. Many of the tests for the push command are redundant (and therefore confusing) assuming good test coverage for the authorization utilities.
3. Many of the tests punch through and exclusively test the private api within the push command.
4. There is some whacky logic around determining the host in the push command that should be cleaned up.

I've got another branch where I attempted to untie the inheritance unsuccessfully (https://github.com/samjonester/rubygems/tree/signin-command-inheritance). It is no small task, and after the attempt I'm not sure it's worth it.

File: test/rubygems/test_gem_commands_signin_command.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    Gem.configuration.rubygems_api_key = nil
+    Gem.configuration.api_keys.clear
+
+    super
+  end
+
+
+
+
+
```
Comment (drbrain) [inline test/rubygems/test_gem_commands_signin_command.rb]: Can you trim this whitespace? (and below)

**Comment by drbrain** [comment]
> Additionally, there are some issues with the push command that should still be addressed. Should I include this as a separate commit in this PR?

Yes.

> Many of the tests for the push command are redundant (and therefore confusing) assuming good test coverage for the authorization utilities.

Yay, deletions!

> Many of the tests punch through and exclusively test the private api within the push command.

That is common throughout the gem command tests since it becomes cumbersome to test details of the implementation by supplying command-line arguments.

> I've got another branch where I attempted to untie the inheritance unsuccessfully (https://github.com/samjonester/rubygems/tree/signin-command-inheritance). It is no small task, and after the attempt I'm not sure it's worth it.

I'll leave some comments on that branch that might make it worth-it.

File: test/rubygems/test_gem_commands_signin_command.rb (hunk 1 above)
Comment (samjonester) [inline test/rubygems/test_gem_commands_signin_command.rb]: Gladly!

**Comment by samjonester** [comment]
Thanks for the feedback, @drbrain! I'll amend the first commit with your suggestions, and get another in for updates to push.

I understand what you mean by being difficult to test the command line arguments. I agree that testing through the `execute` method then seems sufficient. I meant that there was what should have been a private helper method being called by execute and many tests where calling it directly (so that they didn't have to deal with setup for sign in). That was initially why I wanted to untie the inheritance, allowing signin to be faked in the test and also allowing push to be tested through its public API (`execute`).

**Comment by samjonester** [comment]
I've completed a refactor of the push command and welcome feedback / criticisms :) I'm not sure what happened with the travis build. Does anyone have guidance on what I can do to rectify the failure?

File: lib/rubygems.rb [hunk 2]
```diff
@@ -967,6 +967,14 @@ def self.rubygems_version
   end
 
   ##
+  # Is the current rubygems version a beta?
+
+  def self.is_beta?
```
Comment (segiddins) [inline lib/rubygems.rb]: maybe `beta_version?`

Comment (samjonester) [inline lib/rubygems.rb]: sure :) I'd be happy to change that. Thanks for the feedback!

File: test/rubygems/test_gem.rb [hunk 3]
```diff
@@ -29,6 +29,17 @@ def setup
     util_remove_interrupt_command
   end
 
+  def teardown
+    super
+
+    change_version @version if defined? @version
```
Comment (segiddins) [inline test/rubygems/test_gem.rb]: this means calling `change_version` twice won't properly restore the version

Comment (samjonester) [inline test/rubygems/test_gem.rb]: Great point. I felt at a loss of how to "fake this" global state, which is why I landed on this (no ideal) solution. Do you have a suggestions for updating the version without reaching into `Gem` and modifying an instance variable?

File: test/rubygems/test_gem.rb [hunk 4]
```diff
@@ -29,6 +29,17 @@ def setup
...
+  end
+
+  def change_version new_version
+    @version = Gem.instance_variable_get :@rubygems_version
```
Comment (segiddins) [inline test/rubygems/test_gem.rb]: this conflicts with `test/rubygems/test_gem_source_git.rb:10`, maybe use `@gem_orig_version ||=`

File: test/rubygems/test_gem.rb (hunk 3 above)
Comment (segiddins) [inline test/rubygems/test_gem.rb]: using `||=` could work, as I suggested below, else resetting to `Gem::Version.new(Gem::VERSION)`

**Comment by samjonester** [comment]
@segiddins I've incorporated your feedback by amending my commit. Thanks for taking a look! The Travis build is still failing on the same combination in the test matrix, and I'm not sure what the issue is.

**Comment by segiddins** [comment]
Does it pass locally on 1.8.7 for you?

**Comment by samjonester** [comment]
Installing Ruby 1.8.7 correctly (with openssl) took many attempts on OS X El Capitan :crying_cat_face: I know I'm missing the context around the decision and assume there is an important reason, but why is 1.8.7 being supported in future releases of rubygems when it was EOL'd in 2013, then again in 2014? 

Here was approach I took that finally worked for installing 1.8.7

``` sh
brew install ruby-install gcc49
ruby-install --install-dir ~/.rbenv/versions/1.8.7-p375 ruby 1.8.7-p374 -- --without-tk CC=gcc-4.9
rbenv shell 1.8.7-p375
```

Anyways, `rake test` produced the same error locally, which was remedied by changing the `teardown` method to `before_teardown` in `test_gem.rb`, so the PR is passing now :)

Cheers!

**Comment by indirect** [comment]
Thanks for fixing the tests! SemVer means we can't drop 1.8.7 support until we release 3.0. We're planning to do that, but we want to group breaking changes into a single release rather than having many breaking releases.

**Comment by segiddins** [comment]
`gem signin cocoapods` fails for me even though I'm already signed in with an account that can push CocoaPods

**Comment by samjonester** [comment]
I got bit by inheritance again ;) `authorization_utilities` (old `gemcutter_utilities`) and `push`/`signin` defined a method named `host`, and things didn't work unless you manually specified the host with the `--host` flag in the cli. It's sorted out now.

**Comment by segiddins** [comment]
`gem signing cocoapods` still fails

```
ERROR:  While executing gem ... (Gem::Package::FormatError)
    No such file or directory @ rb_sysopen - cocoapods
```

**Comment by samjonester** [comment]
@segiddins Sorry for the inactivity. I've been busy with a move. It's interesting that you're seeing this failure now, as I don't think I changed the behavior of reading the gem by name. Is it possible this was an existing bug? Regardless, I'd be happy to commit a fix, and I'm looking at it now.

Old code: 
``` ruby
gem_data = Gem::Package.new(name)
```

New code:
``` ruby 
def gem_data
  @gem_data ||= Gem::Package.new(get_one_gem_name)
end
```

**Comment by segiddins** [comment]
oooooh, I misunderstood what the command did :/
I thought I could use it to make sure the current user is signed into an account that's authorized to push a gem with the given name

**Comment by samjonester** [comment]
AH! :) No worries. It is designed to perform the initial sign in without actually performing a manual push. It will read the host in a similar way to push, either by providing a flag `--host` or from the gem in the same way as `gem push`, by requiring a gem path like described in the [docs](http://guides.rubygems.org/publishing/#publishing-to-rubygemsorg). I'm not sure that the gem should be required for signing, but wanted to keep the API consistent with `gem push`.

**Comment by segiddins** [comment]
Why not either infer the name from a `.gem` or allow specifying the name via the command line?

**Comment by samjonester** [comment]
You should specify a `.gem`. Here are examples of sister commands for signin and push. Thinking about it again, it does make sense to require signin to provide the gem option, since there could be verification needed between the provided host and the allowed push host.

``` bash
gem signin foo-bar.gem
gem push foo-bar.gem

gem signin --host https://foo.com/bar foo-bar.gem
gem push --host https://foo.com/bar foo-bar.gem
```

**Comment by segiddins** [comment]
Can you document that in the help for the command? Since Im guessing I won't be the only person to think you can just specify the name.

**Comment by samjonester** [comment]
I'd be glad to :) Where else should I include documentation? Where does the source for this page live? http://guides.rubygems.org/publishing/

**Comment by segiddins** [comment]
I was thinking more the output of `--help`, but I think the guides are generated from the guides repo at https://github.com/rubygems/guides

**Comment by samjonester** [comment]
I've updated the help docs, let me know if you would like to change the wording. Also, I've submitted https://github.com/rubygems/guides/pull/161 for the guides.

**Comment by segiddins** [comment]
👍 from me!

**Comment by duckinator** [comment]
This has a lot of unrelated changes and as such it's hard to discern if it works correctly.

I'd like to see it replaced with multiple smaller PRs so they can each be reviewed (and thus merged) more easily:

1. Refactoring the push command (if this is still necessary),
2. Renaming GemcutterUtilities to AuthorizationUtilities (if this is still appropriate),
3. Implementing `gem signin`.

Also: it looks like this implementation of `gem signin` requires you to pass a gemfile to it in order to sign in, unless I'm missing something? This is really weird, especially since `gem push` doesn't &mdash; the current method for signing in is to just do `git push`, which signs you in then throws an error because there's nothing to push. You should be able to just do `gem signin` and authenticate with rubygems.org.

**Comment by duckinator** [comment]
Closing because #1944 replaces this (and was already merged).

**Pull request closed without merging**